build: switch libdispatch build to cmake

compnerd opened the pull request:
Use the cmake based build system for building libdispatch now that it is
able to build Linux, android, and even possible to cross-compile to
Windows.

<!-- What's in this pull request? -->
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

modocache commented:
@swift-ci please test

compnerd commented:
swift-corelibs-libdispatch PR#264 should fix the build issue that happened here

modocache commented:
Please test with following PR:
https://github.com/apple/swift-corelibs-libdispatch/pull/264

@swift-ci Please test

compnerd commented:
CC @dgrove-oss

dgrove-oss commented:
I think we aren't ready to merge this PR yet.  

At least for me, I am seeing the libdispatch tests not linking correctly when built with CMake because of missing -L/-rpath to find libSwift.  I will put in a dispatch PR for that today.

@compnerd  have you build the _entire_ source tree (including swift, swift-corelibs-dispatch, swiftpm, swift-corelibs-foundation) from scratch with CMake driving the libdispatch build with this PR in place for build-script?

compnerd commented:
No, I've not built the entire source tree.  I've been focused on the swift compiler/runtime and libdispatch for now.  I know of at least one other place where we need to tweak things.  But, having an occasional run here is helpful to ensure that I don't accidentally regress the Darwin build.  I've been trying to keep the Darwin build untouched.  Does it not use the autoconf build?  If so, I can greatly simplify this PR.  I know that @das has been working on trying to fix up the CMake build for libdispatch to be sufficient for the macOS build, and I'm really looking forward to that to have a single unified build for libdispatch.

compnerd committed:
build: switch libdispatch build to cmake

Use the cmake based build system for building libdispatch now that it is
able to build Linux, android, and even possible to cross-compile to
Windows.

compnerd commented:
@swift-ci please test

Lets see where we stand

gottesmm commented:
(Just doing a drive by test since the CI didn't seem to kick in).

gottesmm commented:
@swift-ci please test

gottesmm commented:
@swift-ci please test

compnerd commented:
Interesting, seems that the linker invocation is missing the `-rpath-link` for the swift runtime.

dgrove-oss commented:
I think the missing -rpath-link is because on the dispatch side it is looking for SWIFT_RUNTIME_LIBDIR to be defined.

dgrove-oss commented on utils/build-script-impl:
```diff
@@ -2443,41 +2444,60 @@ for host in "${ALL_HOSTS[@]}"; do
+                    -DCMAKE_CXX_COMPILER:PATH="${LLVM_BIN}/clang++"
+                    -DCMAKE_SWIFT_COMPILER:PATH="${SWIFTC_BIN}"
+                    -DCMAKE_INSTALL_PREFIX:PATH="$(get_host_install_destdir ${host})$(get_host_install_prefix ${host})"
+                    -DCMAKE_RUNTIME_LIBDIR:PATH="${DISPATCH_TMPLIBDIR}"
```
SWIFT_RUNTIME_LIBDIR instead I think

compnerd commented:
Rebased and adjusted as per @dgrove-oss's point.

compnerd commented:
@swift-ci please test

compnerd commented:
Switched the sourcekit's build to use cmake as well.  Adjusted the build-script to fix the rpath for swiftcore

compnerd commented:
@swift-ci please test Linux platform

compnerd commented:
@swift-ci please test

compnerd commented:
Okay, so, the last failure was an install failure.  I think that was caused by me using `${CMAKE_INSTALL_PREFIX}` for the install prefix.  Switch that to `<INSTALL_DIR>` which is the temporary install location for the `ExternalProject`.  Hopefully, this should finally get it right.

compnerd commented:
@swift-ci please test

compnerd commented:
Okay, swift-corelibs-libdispatch PR303 should fix the symlink issue.  It seems that I had them locally which is why I never saw this.  I adjusted the dependencies to deal with BlocksRuntime dependencies being needed as well.

compnerd commented:
Please test with following pull request:
https://github.com/apple/swift-corelibs-libdispatch/pull/303

@swift-ci Please test Linux platform

CodaFi commented:
Please test with following pull request:
apple/swift-corelibs-libdispatch#303

@swift-ci Please test Linux platform

compnerd commented:
@CodaFi thanks!  Great, it now seems to also pass on the build bots!  This is great!  Need to wait for @das to get the PR#303 merged in libdispatch and then we can switch over to the CMake build! ❤️

compnerd commented:
@swift-ci please smoke test macOS platform

compnerd commented:
Please test with following pull request:
apple/swift-corelibs-libdispatch#303

@swift-ci Please test macOS platform

compnerd commented:
Just need to wait for the dispatch PR to get merged before this can be merged.

gottesmm commented:
@compnerd I am soooo excited about this! Getting rid of autoconf from the build? +1! +1!

gottesmm commented:
Woot!
